tracer/src/Datadog.Trace/Headers/Ip/IpExtractor.cs

```diff
@@ -0,0 +1,141 @@
+
+                var success = IPAddress.TryParse(consideredPotentialIp, out var ipAddress);
+                if (success)
+                {
+                    if (ipAddress.IsIPv4MappedToIPv6)
+                    {
+                        ipAddress = ipAddress.MapToIPv4();
+                    }
+
+                    if (IsPrivateIp(ipAddress))
+                    {
+                        // only set the oldest (so the first one)
+                        privateIpInfo ??= addressAndPort;
+                    }
+                    else
+                    {
+                        var publicIpInfo = new IpInfo(ipAddress.ToString(), addressAndPort.Port);
+                        return publicIpInfo;
+                    }
+                }
+            }
+
+            return privateIpInfo;
+        }
+
+        internal static IpInfo ExtractAddressAndPort(string ip, bool https = false, int? defaultPort = null)
+        {
+            var port = defaultPort ?? DefaultPort(https);
+            if (ip.Contains("."))
+            {
+                var parts = ip.Split(':');
+                if (parts.Length == 2)
+                {
+                    _ = int.TryParse(parts[1], out port);
+                    return new IpInfo(parts[0], port);
+                }
+            }
+
+            var result = _ipv6Regex.Match(ip);
+            if (result.Success)
+            {
+                ip = result.Groups[1].Captures[0].Value;
+                _ = int.TryParse(result.Groups[2].Captures[0].Value, out port);
+            }
+
+            return new IpInfo(ip, port);
+        }
+
+        internal static bool IsPrivateIp(IPAddress ipAddress)
+        {
+            if (ipAddress.AddressFamily != System.Net.Sockets.AddressFamily.InterNetworkV6)
+            {
+                var ipAddr = BitConverter.ToInt32(ipAddress.GetAddressBytes(), 0);
+                foreach (var currentCidrMask in _ipv4LocalCidrs)
+                {
+                    var result = (ipAddr & currentCidrMask.Item2) == (currentCidrMask.Item1 & currentCidrMask.Item2);
+                    if (result)
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            if (ipAddress.IsIPv6SiteLocal || ipAddress.IsIPv6LinkLocal)
+            {
+                return true;
+            }
+
+#if NET6_0_OR_GREATER
+            return ipAddress.IsIPv6UniqueLocal;
+#else
+            var firstWord = ipAddress.ToString().Split(new[] { ':' }, StringSplitOptions.RemoveEmptyEntries)[0];
+            // These days Unique Local Addresses (ULA) are used in place of Site Local. ULA has two variants:
+            // fc00::/8 is not defined yet, but might be used in the future for internal-use addresses
+            // fd00::/8 is in use and does not have to registered anywhere.
+            if (firstWord.Length >= 4)
+            {
+                var first2Letters = firstWord.Substring(0, 2);
```

Comment:
nit: Super small, but instead of using `string.Substring` which may allocate, you could compare the characters at index 0 and index 1